storage/log_storage.go

```diff
@@ -116,6 +116,15 @@ type LogStorage interface {
 	// the returned object, and values read through it should only be propagated
 	// if Commit returns without error.
 	BeginForTree(ctx context.Context, treeID int64) (LogTreeTX, error)
+
+	// QueueLeaves enqueues leaves for later integration into the tree.
+	// If error is nil, the returned slice of leaves will be the same size as the
+	// input, and each entry will hold:
+	//  - the existing leaf entry if a duplicate has been submitted
+	//  - nil otherwise.
+	// Duplicates are only reported if the underlying tree does not permit duplicates, and are
+	// considered duplicate if their leaf.LeafIdentityHash matches.
+	QueueLeaves(ctx context.Context, treeID int64, leaves []*trillian.LogLeaf, queueTimestamp time.Time) ([]*trillian.LogLeaf, error)
```

Comment:
Since this method is now allowed to fail queuing some of the entries, don't we need to return some indication of that? E.g., return `QueuedLogLeaf` instead of `LogLeaf` to associate more elaborate statuses with the entries.

Reply:
You might be right, but I'll leave that for another PR to keep this as small as possible.

Reply:
Yes - done